File: sentimental-credit/credit.py
# TODO
import re


def main():
    number = get_number()
    if check_number(number):
        check_card(number)
    else:
        print("INVALID")


def get_number():
    number = input("Number: ")
    return number


def check_number(number):
    list = [int(x) for x in number]
    sum = 0
    for i in range(-2, -len(number) - 1, -2):
        if list[i] * 2 >= 10:
            j = list[i] * 2
            j = j % 10 + 1
            sum += j
        else:
            sum += list[i] * 2
    for i in range(-1, -len(number) - 1, -2):
        sum += list[i]
    if sum % 10 == 0:
        return True
    else:
        return False


def check_card(number):
    amex = r"^(34|37)\d{13}$"
    mastercard = r"^5[1-5]\d{14}$"
    visa = r"^4(\d{12}|\d{15})"
    if re.fullmatch(amex, number):
        print("AMEX")
    elif re.fullmatch(mastercard, number):
        print("MASTERCARD")
    elif re.fullmatch(visa, number):
        print("VISA")
    else:
        print("INVALID")


if __name__ == "__main__":
    main()

File: credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

int add_digits(long number, long i, int multiple);
string checksum(long number);
string checkcard(long number);

int main(void)
{
    // get card number
    long number = get_long("Number: ");
    //check number if it is a valid number, if so which credit card is it
    printf("%s", checksum(number));
}

string checksum(long number)
{
    //add_digits function to add digits with the last argument being the multiplier and the second argument
    int sum = add_digits(number, 100, 2) + add_digits(number, 10, 1);
    if (sum % 10 == 0)
    {
        return checkcard(number);
    }
    else
    {
        return ("INVALID\n");
    }

}

int add_digits(long number, long last, int multiple)
{
    int remainder = 0;
    int sum = 0;
    for (long x = last; x <= number * 10; x *= 100)
    {
        long j = x / 10;
        remainder = ((number % x) / j) * multiple;
        if (remainder >= 10)
        {
            remainder = remainder % 10 + remainder / 10;
        }
        sum += remainder;
    }
    return sum;
}

string checkcard(long number)
{
    if ((((long)(pow(10, 13) * 35)) <= number && number < ((long)(pow(10, 13) * 36))) || (((long)(pow(10, 13) * 37)) <=  number
            && number < ((long)(pow(10, 13) * 38))))
    {
        return ("AMEX\n");
    }
    else if ((long)(pow(10, 14) * 51) <=  number && number < (long)(pow(10, 14) * 56))
    {
        return ("MASTERCARD\n");
    }
    else if ((((long)(pow(10, 12) * 4)) <=  number && number < ((long)(pow(10, 12) * 5))) || (((long)(pow(10, 15) * 4)) <=  number
             && number < ((long)(pow(10, 15) * 5))))
    {
        return ("VISA\n");
    }
    else
    {
        return ("INVALID\n");
    }
}